### ISSUE TITLE
Support new `eslint.config.js` file

### ISSUE THREAD

**Issue by b0o** [mantoni/eslint_d.js#214]
ESLint now has a [new `eslint.config.js` configuration file](https://eslint.org/docs/latest/user-guide/configuring/configuration-files-new).

Running `eslint` works as expected, but `eslint_d` doesn't find the configuration file:

```
$ eslint -f tap .
TAP version 13
1..2
ok 1 - /path/to/my/project/eslint.config.js
ok 2 - /path/to/my/project/index.js

$ eslint_d -f tap .
Error: No ESLint configuration found in /path/to/my/project.

$ ls -la
drwxr-xr-x 1 user user   416 Sep 13 16:57 .
drwxr-xr-x 1 user user  1598 Sep 13 15:32 ..
drwxr-xr-x 1 user user   164 Sep 13 14:47 assets
-rw-r--r-- 1 user user   754 Sep 13 16:54 eslint.config.js
drwxr-xr-x 1 user user   178 Sep 13 16:14 .git
-rw-r--r-- 1 user user  4175 Sep 13 15:28 .gitignore
-rw-r--r-- 1 user user  1536 Sep 13 16:33 index.js
-rw-r--r-- 1 user user  1075 Sep 12 21:40 LICENSE
-rw-r--r-- 1 user user   513 Sep 13 15:19 manifest.json
drwxr-xr-x 1 user user  2312 Sep 13 16:10 node_modules
-rw-r--r-- 1 user user   892 Sep 13 16:10 package.json
-rw-r--r-- 1 user user 84034 Sep 13 16:10 package-lock.json
-rw-r--r-- 1 user user    20 Sep 13 15:36 .prettierrc.json
-rw-r--r-- 1 user user  3246 Sep 13 15:28 README.md
```

(I'm using the `tap` formatter here just to get some output when linting has no errors)

**Comment by mantoni** [comment]
Thank you for opening and for the link to the docs.

This is an experimental feature. From digging into the ESLint PRs, they have some temporary hacks to enable the new configs if the file exists. However this is going to change once they completely switch to the `FlatESLint` class.

We'd have to replicate these checks in `eslint_d` and then remove them once the feature fully landed. Since `eslint_d` can work with multiple `eslint` versions, we'd also have to feature detect / version check the `eslint` version being used and apply the hacks accordingly.

I'm a little afraid that this is overkill and might even cause stability problems with `eslint_d`. I'd rather wait until the new implementation becomes the default with the next major release of `eslint`.

Does that make sense?

**Comment by b0o** [comment]
Sounds good to me, thank you for looking into this.

**Comment by tlvince** [comment]
Apparently this format has stablised and will be the default in eslint v9 and is [currently opt-in](https://eslint.org/docs/latest/use/configure/configuration-files-new).

**Comment by mantoni** [comment]
@tlvince Thank you for the update.

I'm currently a little under water with work, so won't have time to implement this myself. If anyone wants to take a look, I'm happy to assist and review a PR. ❤️

**Comment by amritk** [comment]
hey @mantoni I would be willing to help out here if you can point me in the right direction. I switched back to eslint and it is unbearably slow. Do we just need to look for the new config file? And then maybe migrate everything over to ESM?

**Comment by mantoni** [comment]
@amritk Awesome!

- So first of all we should use the same opt-in behaviour as eslint, which is checking for `ESLINT_USE_FLAT_CONFIG`.
- If I understand [the docs](https://eslint.org/docs/latest/use/configure/configuration-files-new#configuration-file) correctly, we have to use the `FlatESLint` class if the environment variable is defined.
- Have a look here, where we invoke `executeWithESLint `:

https://github.com/mantoni/eslint_d.js/blob/2681a31138997776fcaa8944c8af2853cbb8dc36/lib/linter.js#L310

Maybe all we need to do is pass in the `FlatESLint` class? If that doesn't work we might need a new function `executeWithFlatESLint`? Something to find out. I guess looking at how eslint implements the switch itself might help. 

I don't see a reason why an ESM migration is coupled / required for this. Can you enlighten me?

**Comment by amritk** [comment]
@mantoni ur right we don't need to migrate to ESM, but we do need to change wherever the config file is imported 
```ts
Error [ERR_REQUIRE_ESM]: Cannot read config file: /home/amritk/apps/ak/portfolio/eslint.config.js
Error: require() of ES Module /home/amritk/apps/ak/portfolio/eslint.config.js from /home/amritk/apps/pulls/eslint_d.js/node_modules/@eslint/eslintrc/dist/eslintrc.cjs not supported.
Instead change the require of eslint.config.js in /home/amritk/apps/pulls/eslint_d.js/node_modules/@eslint/eslintrc/dist/eslintrc.cjs to a dynamic import() which is available in all CommonJS modules.
```

Also anything I log in the linter.js file doesn't show up, which makes me think its blowing up before that somewhere.

I was trying to track it down but the last log I can see before the error is in `core_d`
https://github.com/mantoni/core_d.js/blob/master/lib/client.js#L77

Then I'm not quite sure where it goes and blows up

**Comment by mantoni** [comment]
Yeah, regarding the error logs, they're probably going nowhere (see #87). I don't have anything handy apart from unit tests, sorry. You could write stuff to a file in the cwd as a hack.

Regarding module configs, the eslint documentation states this:

> If your project does not specify "type":"module" in its package.json file, then eslint.config.js must be in CommonJS format ...

So `eslint_d` should do the same and use dynamic import if `type` is `module` and require otherwise.

**Comment by amritk** [comment]
>So eslint_d should do the same and use dynamic import if type is module and require otherwise.
Yep exactly, just gotta find out where the import occurs. Maybe I'll add the log option as well

**Comment by amritk** [comment]
Alright so update:
I sorted out some logs and was able to track it down. I was testing out using the FlatESLint class but it looks like it takes some different options
```ts
Error: Invalid Options:
- Unknown options: extensions, ignorePath, resolvePluginsRelativeTo, rulePaths, useEslintrc
- 'extensions' has been removed.
- 'resolvePluginsRelativeTo' has been removed.
- 'ignorePath' has been removed.
- 'rulePaths' has been removed. Please define your rules using plugins.
```

Should be able to figure this out when I get some time next week!

**Comment by amritk** [comment]
alright @mantoni  I gave it another go. I got all the options mostly working now for the flat config. The issue now is actually resolving the flatconfig class instead of the eslint one. For some reason it keeps looking in `eslint_d.js/node_modules/eslint/lib/api.js` but we need to pull the class from `eslint_d.js/node_modules/eslint/lib/eslint` I'm not sure where this is done, could you help me out?

**Comment by mantoni** [comment]
Awesome! I had a brief look at the eslint source, and they have this require statement in their cli:

```js
{ FlatESLint, shouldUseFlatConfig } = require("./eslint/flat-eslint")
```

So I guess you will have to explicitly require `${absolute_eslint_path}/lib/eslint/flat-eslint` in `caches.js` here:

https://github.com/mantoni/eslint_d.js/blob/2681a31138997776fcaa8944c8af2853cbb8dc36/lib/caches.js#L48C21-L48C41

**Comment by amritk** [comment]
I found it in unsupportedAPI
https://github.com/eslint/eslint/blob/main/lib/unsupported-api.js
So I guess I'll pass down a flag for flatconfig to caches and just replace `/lib/api.js` with `/lib/unsupported-api.js`
That should work right?

**Comment by mantoni** [comment]
This feature was implemented in #277 and released in `v13.1.0` 🎉
